iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/WeakHashAlgorithmsCheck.java

```diff
@@ -0,0 +1,164 @@
+import org.sonar.iac.common.api.tree.impl.TextRanges;
+import org.sonar.iac.docker.symbols.ArgumentResolution;
+import org.sonar.iac.docker.tree.api.RunInstruction;
+
+@Rule(key = "S4790")
+public class WeakHashAlgorithmsCheck implements IacCheck {
+
+  private static final String MESSAGE = "Using weak hashing algorithms is security-sensitive.";
+  private static final Set<String> OPENSSL_SENSITIVE_SUBCOMMAND = Set.of("md5", "sha1", "rmd160", "ripemd160");
+  private static final Set<String> OPENSSL_SENSITIVE_DGST_OPTION  = Set.of("-md2", "-md4", "-md5", "-sha1", "-ripemd160", "-ripemd", "-rmd160");
+
+  private static final CommandDetector sensitiveOpenSslSubcommand = CommandDetector.builder()
+    .with("openssl"::equals)
+    .with(OPENSSL_SENSITIVE_SUBCOMMAND::contains)
+    .build();
+  private static final CommandDetector sensitiveOpenSslDgst = CommandDetector.builder()
+    .with("openssl"::equals)
+    .with("dgst"::equals)
+    .withOptional(s -> !OPENSSL_SENSITIVE_DGST_OPTION.contains(s) && s.startsWith("-"))
+    .with(OPENSSL_SENSITIVE_DGST_OPTION::contains)
+    .build();
+
+  @Override
+  public void initialize(InitContext init) {
+    init.register(RunInstruction.class, WeakHashAlgorithmsCheck::checkRun);
+  }
+
+  private static void checkRun(CheckContext ctx, RunInstruction runInstruction) {
+    List<ArgumentResolution> resolvedArgument = runInstruction.arguments().stream().map(ArgumentResolution::of).collect(Collectors.toList());
+
+    sensitiveOpenSslSubcommand.search(resolvedArgument).forEach(command -> ctx.reportIssue(command, MESSAGE));
+    sensitiveOpenSslDgst.search(resolvedArgument).forEach(command -> ctx.reportIssue(command, MESSAGE));
+  }
+
+  static class Command implements HasTextRange {
+    List<ArgumentResolution> resolvedArguments;
+
+    public Command(List<ArgumentResolution> resolvedArguments) {
+      this.resolvedArguments = resolvedArguments;
+    }
+
+    @Override
+    public TextRange textRange() {
+      return TextRanges.mergeElementsWithTextRange(resolvedArguments.stream().map(ArgumentResolution::argument).collect(Collectors.toList()));
+    }
+  }
+
+  private static class CommandDetector {
+
+    List<CommandPredicate> predicates;
+
+    private CommandDetector(List<CommandPredicate> predicates) {
+      this.predicates = predicates;
+    }
+
+    public static Builder builder() {
+      return new Builder();
+    }
+
+    /**
+     * Return all block of arguments which match the list of predicates.
+     */
+    private List<Command> search(List<ArgumentResolution> resolvedArguments) {
+      List<Command> commands = new ArrayList<>();
+
+      for (int argIndex = 0; argIndex < resolvedArguments.size(); argIndex++) {
+        Integer sizeCommand = fullMatch(resolvedArguments, argIndex);
+        if (sizeCommand != null) {
+          commands.add(new Command(resolvedArguments.subList(argIndex, argIndex + sizeCommand)));
+          argIndex += sizeCommand - 1;
```

Comment:
This is still not covered by a test.
`argIndex += sizeCommand - 2;` does not lead to any failing test.

Reply:
Good point. New tests are covering the previous modification which was something like `argIndex += sizeCommand + 2;`, but not this modification with `- 2`.
This one is a bit more tricky as `argIndex += sizeCommand - 1;` is not required to make it work, it's more like an optimization to skip the full matched command. We would need a sensitive command which give us a sizeCommand of 1 to ensure this, which we don't have for now with OpenSSL (2+ arguments).

I created some specific test for CommandDetector class, for now it is in the class WeakHashAlgorithmsCheckTest, but once we extract CommandDetector, we should create a dedicated test class for it and move those test in it. I added a todo about this.